rfcs/taskforce/0003-Integration-test-infrastructure-spec.md

```diff
@@ -0,0 +1,204 @@
+
+In the current testnet infrastructure, a network manager would generate the necessary terraform configuration files and deploy using terraform.
+To run it locally, using minikube perhaps, we would create another network manager that deploys a testnet locally.
+
+The network manager itself can be specified as:
+
+```ocaml
+
+module Node : Node_intf
+(* Deploy the network*)
+val deploy : Network_config.t -> Daemon_config.t -> Testnet.t Deferred.t
+
+(*Tear down the network*)
+val destroy : Testnet.t -> unit Deferred.t
+```
+
+```ocaml
+
+module type Node_intf : sig
+    type t
+
+    val node_id_json : t -> json
+
+    val start : t -> unit Deferred.t
+
+    val stop : t -> unit Deferred.t
+
+    ...
+end
+
+module type Testnet = sig
+
+    type t =
+        {
+          block_producers : Node.t list
+        ; snark_coordinators: Node.t list
+        ; archive_nodes: Node.t list
+        ; ... }
+end
+```
+
+A node will encapsulate all the information needed to connect to it and make requests such as sending payments, retrieving a node's status, and so on. These requests implemented by the network manager could be either using graphql commands or cli commands. The interface should also implement functions to start and stop a node that can be triggered at any time by the test program. This is particularly needed for testing functionalities like bootstrap and catchup. The network infrastructure should facilitate stopping and restarting coda daemon containers (with or w/o persisting the config directory). Currently a pod or a container within a pod is restarted automatically every time the daemon terminates. This behavior should be changed to support restarting a node on request. The following are some of the options to enable restart on request:
+    1. Have another process as the root of the container that initiates or terminates coda daemon based on a request from the network manager. This requires implementing and maintaining another application.
+    2. Using kubernetes init containers to start/stop coda daemon containers. The communication between the network manager and the init containers will have to be facilitated by incorporating some sort of message queues in the deployment architecture. This would not require a custom application that needs to be maintained but adds more complexity to the deployment architecture.
```

Comment:
I don't think this will actually work. As far as I understand, init containers are run with a container is provisioned, not when a container restarts. The init containers are run once when the container is first booted up on a pod, so they won't rerun when the container dies unless the pod is also evicted and a new one is created in it's place. This also seems much more complex than a puppeteer process in the main container. If you have a puppeteer in the main container, then you can always expect to be able to talk to it so long as the container is running, regardless of the state of the node on the container. If we use init containers, the executive needs to be aware of the state of the node to know whether or not to tell the node to stop (daemon graphql) or to tell the waiting process to start the node. That state can drift in the event of crashes or pod evictions, so it can't be guaranteed to be in sync, and would require some fault tolerance on the executive's part to handle that condition.